cli/command/telemetry_utils.go

```diff
@@ -0,0 +1,156 @@
+package command
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type contextKey string
+
+const (
+	OtelMeterCtxKey  = contextKey("otel.meter")
+	OtelTracerCtxKey = contextKey("otel.tracer")
+)
+
+// -------- Some utils, mostly adapted from compose-cli-private and buildx for testing purposes --------
+
+// BaseMetricAttributes returns an attribute.Set containing attributes to attach to metrics/traces
+func BaseMetricAttributes(cmd *cobra.Command) attribute.Set {
+	attrList := []attribute.KeyValue{
+		attribute.String("command.name", getFullCommandName(cmd)),
+	}
+	return attribute.NewSet(attrList...)
+}
+
+// InstrumentCobraCommands wraps all cobra commands' RunE funcs to set a command duration metric using otel.
+//
+// Note: this should be the last func to wrap/modify the PersistentRunE/RunE funcs before command execution.
+//
+// can also be used for spans!
+func InstrumentCobraCommands(ctx context.Context, cmd *cobra.Command, meter metric.Meter) {
+	// If PersistentPreRunE is nil, make it execute PersistentPreRun and return nil by default
+	ogPersistentPreRunE := cmd.PersistentPreRunE
+	if ogPersistentPreRunE == nil {
+		ogPersistentPreRun := cmd.PersistentPreRun
+		ogPersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+			ogPersistentPreRun(cmd, args)
+			return nil
+		}
+		cmd.PersistentPreRun = nil
+	}
+
+	// wrap RunE in PersistentPreRunE so that this operation gets executed on all children commands
+	cmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+		// If RunE is nil, make it execute Run and return nil by default
+		ogRunE := cmd.RunE
+		if ogRunE == nil {
+			ogRun := cmd.Run
+			ogRunE = func(cmd *cobra.Command, args []string) error {
+				ogRun(cmd, args)
+				return nil
+			}
+			cmd.Run = nil
+		}
+		cmd.RunE = func(cmd *cobra.Command, args []string) error {
+			// insert the meter and tracer objects into the context
+			// so commands can use them in their RunE func
+			newCtx := context.WithValue(cmd.Context(), OtelMeterCtxKey, meter)
+			// newCtx := context.WithValue(newCtx, OtelTracerCtxKey, tracer))
+			cmd.SetContext(newCtx)
```

Comment:
@laurazard @Benehiko 

As an extra little test thing we can discuss,  I added the otel `Meter` to the cobra command context as a value, so that commands can retrieve it and use it in their RunE funcs (for example for adding command specific data). We can then do the same thing for tracing if it's useful. What do you think?

Reply:
I think it depends, using `ctx` to store values could make using `meter` clunky. Also adding it here to the cmd ctx inside the `RunE/Run` function might not be necessary. You can add it way earlier where you create the meter instance to the top level context and execute cobra with the `ExecuteContext` method https://github.com/spf13/cobra/blob/v1.8.0/command.go#L1030

e.g. `meter, ok := ctx.Value(OtelMeterCtxKey).(metric.Meter)`

Just a thought but couldn't we just add the meter to the dockerCli instance and have a nice method for using it? `dockerCli.Meter()`

Reply:
To expand on the `ctx` approach, we could also create standalone utility functions which help with using the meter,

```go
package utils

func Meter(ctx context.Context) (metric.Meter, bool) {
  return ctx.Value(OtelMeterCtxKey).(metric.Meter)
}
```

```go
meter, ok := utils.Meter(ctx)
if !ok {
  return
}
```

Reply:
> You can add it way earlier

So my thinking was to only include it in the `RunE` context, and to do so when instrumenting the cobra command so most of these otel things happened in the same place, so it seemed like a good place to put it for now. No hard feelings here and I can move it to wherever we think is best (or even remove it altogether).

> Just a thought but couldn't we just add the meter to the dockerCli instance and have a nice method for using it? dockerCli.Meter()

I briefly thought about it, but I'd really prefer to avoid adding more things to that struct unless we're totally sure it's the right place (also since the cli itself - including that struct - is used in other projects as well). 
I'm actually slightly more intrigued by the idea of moving some data from dockerCli to the context, but that for a whole other discussion :) 

> To expand on the ctx approach, we could also create standalone utility functions which help with using the meter

Totally agree with this! I left a util out just for the sake of discussing the approach first

Reply:
> I briefly thought about it, but I'd really prefer to avoid adding more things to that struct unless we're totally sure it's the right place (also since the cli itself - including that struct - is used in other projects as well).
I'm actually slightly more intrigued by the idea of moving some data from dockerCli to the context, but that's for a whole other discussion :)

Good point!  I'm not a big fan of values inside context just for the sake of it. To be honest I would actually want a singleton that contains all of these init instances necessary for the CLI to function (like otel) and then expose the utility of them through an interface that's passed around. Then you don't need to use `ctx` with lots of utility functions.

```go
package main

import "context"

type Singleton interface {
	command.Cli
	Meter()
}

type singleton struct {
	meter metric.Meter
	dockerCli command.DockerCli
}

func (s *singleton) DoSomething() {
}

func main() {
	s := &singleton{
		meter: metric.NewMeter(),
		dockerCli: command.NewDockerCli(),
	}
	NeedSingleton(ctx, s)
}

func NeedSingleton(ctx context.Context, s Singleton) {
	s.Meter()
	s.Client()
}

func NeedPartialSingleton(ctx context.Context, s interface {
	Meter()
}) {
	s.Meter()
}
```

Reply:
I think that was _kinda_ the reasoning behind the `DockerCli` struct itself. I'm not sure this added complexity is something we should deal with right now, and I'd focus on getting the minimum functionality we need (so whatever metrics the compose wrapper currently handles, so we can remove it). I don't mind the suggested approach, so let's keep it in mind if others agree.

Even adding the meter to the context is essentially overkill right now, as it's not actually needed for anything (but will eventually be once we want to measure things during the execution of commands)

Reply:
oh yeah i agree, now isn't the right time for this :)
I'd be fine to keep the meter in the context for now if it means we an ship this.

Reply:
I would avoid this context stuff in general. My recommendation is to use [otel.SetTracerProvider](https://pkg.go.dev/go.opentelemetry.io/otel#SetTracerProvider) and [otel.SetMeterProvider](https://pkg.go.dev/go.opentelemetry.io/otel#SetMeterProvider) and then use the corresponding function in code that uses these providers.

For traces, you also pretty rarely need to access the global tracer because the original `TracerProvider` is accessible from [trace.Span](https://pkg.go.dev/go.opentelemetry.io/otel/trace#Span).

A useful pattern that I found is creating a utility similar to this one: https://github.com/docker/buildx/blob/fe08cf2981836f1828058f02cb7684f676b53a9e/util/metricutil/metric.go#L216-L219

In general, you should not pass or keep a meter and instead just the meter provider.

Reply:
I'm not super sure about this. I agree with the meter provider being set with `otel.SetMeterProvider` and will change that, but reading the otel documentation it seems like it'd be an error to create multiple meters with the same "identity", which is why I added the pre-created meter into the ctx.

Maybe I am misunderstanding something? 

Snippet from [the docs](https://opentelemetry.io/docs/specs/otel/metrics/api/#meterprovider)

> Meters are identified by name, version, and schema_url fields. When more than one Meter of the same name, version, and schema_url is created, it is unspecified whether or under which conditions the same or different Meter instances are returned. **It is a user error to create Meters with different attributes but the same identity.**

Reply:
> It is a user error to create Meters with different attributes but the same identity.

The docs specify that it would be error to have different meters with the same identity **but different attributes**, but in this case they would have the same attributes right? So I think this would be fine @krissetto.

Reply:
I think this depends. When adding the MeterProvider instance to ctx instead of the Meter instance, it gives opportunity inside the code to create more meters with unique package names. 

>A Meter should be scoped at most to a single package.
https://pkg.go.dev/go.opentelemetry.io/otel/metric#MeterProvider

So as far as I understand this, a new meter would be created per package and from there you would create a new instrument to do things like count etc. https://opentelemetry.io/docs/specs/otel/metrics/api/#meter-operations

From my understanding that would make @jsternberg correct. But this would mean we would need to detect for meter name collisions (unless this is already baked into the Go sdk).

Reply:
edit: also that would mean we might not need to pass MeterProvider to ctx.

Reply:
> > It is a user error to create Meters with different attributes but the same identity.
> 
> The docs specify that it would be error to have different meters with the same identity **but different attributes**, but in this case they would have the same attributes right? So I think this would be fine @krissetto.

Mmh i'm not totally sure about this. If otel implicitly returns the same Meter when we give the MeterProvider the same name/version, then I totally agree with the approach. The documentation is unclear with this though, so I will be doing some tests and changing the impl if the meter returned is the same :)

Reply:
> A useful pattern that I found is creating a utility similar to this one: https://github.com/docker/buildx/blob/fe08cf2981836f1828058f02cb7684f676b53a9e/util/metricutil/metric.go#L216-L219

I called this `GetDefaultMeter` for now, to avoid confusion with the pre-existing `metric.MeterProvider.Meter()` func

Reply:
> Mmh i'm not totally sure about this. If otel implicitly returns the same Meter when we give the MeterProvider the same name/version, then I totally agree with the approach. The documentation is unclear with this though, so I will be doing some tests and changing the impl if the meter returned is the same :)

Yea, it returns the same meter if you give it the same name/version. If you have an area that calls this repeatedly, then it can be useful to pass around the meter, but it's my opinion that `MeterProvider` should be used as the external API for most if not all situations.

It also mentions that it's an error to put different attributes on the meter, but it's not actually possible to put attributes on the meter anyway in the current version of the SDK. There's an API for instrumentation attributes (which is the name of the type of attribute that you send to the meter) but they're not hooked up anywhere.

> I think this depends. When adding the MeterProvider instance to ctx instead of the Meter instance, it gives opportunity inside the code to create more meters with unique package names.
>
> A Meter should be scoped at most to a single package.
https://pkg.go.dev/go.opentelemetry.io/otel/metric#MeterProvider
>
> So as far as I understand this, a new meter would be created per package and from there you would create a new instrument to do things like count etc. https://opentelemetry.io/docs/specs/otel/metrics/api/#meter-operations

This is true, but my recommendation is to scope the meter to the go module path rather than the package name. It ends up being a bit too difficult to have a different meter per package when processing them downstream. Per-module is a good middle ground in my opinion and it's what we're going to do in buildx for now.

If I'm wrong about this, we can update it later and scope it further. It's mostly about avoiding conflicts with metric names and it should be possible to avoid conflicts within the same module.